File: core/src/main/java/net/siegerpg/siege/core/party/Party.java
package net.siegerpg.siege.core.party;

import lombok.Getter;
import net.siegerpg.siege.core.Core;
import org.bukkit.Bukkit;
import org.bukkit.OfflinePlayer;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.entity.Player;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.UUID;

@SuppressWarnings("ALL")
public class Party {
    @Getter private final UUID partyID = UUID.randomUUID();
    @Getter private final List<UUID> invited = new ArrayList<>();
    private HashSet<UUID> members = new HashSet<>();
    private UUID leader = null;

    /**
     * Create a new party
     *
     * @param leader The leader of the party
     */
    public Party(Player leader) {
        this.setLeader(leader.getUniqueId());
        save();
    }

    /**
     * Create a new party
     *
     * @param partyID The uuid of the party, used to retrieve the party from the config.
     */
    public Party(UUID partyID) {
        //  FileConfiguration config = Core.plugin().partyConfig.getConfiguration(); TODO: Uncomment when adding parties back
//        String leader = config.getString("party." + partyID.toString() + ".leader");
//        List<String> members = config.getStringList("party." + partyID.toString() + ".members");
//        if (leader == null) {
//            Core.INSTANCE.getLogger().warning("Failed to fetch data for party ID " + partyID.toString());
//            return;
//        }

//        UUID partyLeader = UUID.fromString(leader);
//        List<UUID> membersConverted = new ArrayList<>();
//        members.forEach((String u) -> membersConverted.add(UUID.fromString(u)));
//
//        this.partyID = partyID;
//        this.setLeader(partyLeader);
//        for (UUID uuid : membersConverted) addMember(uuid);
    }

    /**
     * Get the leader of the party
     *
     * @return OfflinePlayer, the leader
     */
    public OfflinePlayer getLeader() {
        return Bukkit.getOfflinePlayer(leader);
    }

    /**
     * Add a member to the party
     *
     * @param memberUUID The uuid of the player
     */
    public void addMember(UUID memberUUID) {
        members.add(memberUUID);
        save();
    }

    /**
     * Remove a member from the party
     *
     * @param memberUUID The uuid of the player
     */
    public void removeMember(UUID memberUUID) {
        members.remove(memberUUID);
        save();
    }

    /**
     * Get the members of the party
     *
     * @return The members AND the leader
     */
    public ArrayList<OfflinePlayer> getMembers() {
        ArrayList<OfflinePlayer> list = new ArrayList<>();
        list.add(getLeader());
        for (UUID uuid : members) list.add(Bukkit.getOfflinePlayer(uuid));
        return list;
    }

    /**
     * Get the uuids of the members of the party
     *
     * @return The uuids of the members AND the leader
     */
    public List<UUID> getMembersRaw() {
        List<UUID> list = new ArrayList<>();
        list.add(leader);
        list.addAll(members);
        return list;
    }

    /**
     * See whether or not an UUID is apart of the members
     *
     * @param playerUUID
     * @return
     */
    public boolean isMember(UUID playerUUID) {
        return getMembersRaw().contains(playerUUID);
    }

    /**
     * Add a player to the list of pending invitees (and automatically remove them after 1200 ticks if they haven't joined)
     *
     * @param invitee
     */
    public void addInvite(Player invitee) {
        invited.add(invitee.getUniqueId());
        new BukkitRunnable() {
            @Override
            public void run() {
                if (!isInvited(invitee)) return;
                removeInvite(invitee);
                send(invitee.getName() + " did not join the party!");
                invitee.sendMessage("§cParty invite from " + getLeader().getName() + " has expired!");
                if (members.size() == 0) {
                    send("The party was disbanded since there was only 1 person left!");
                    disband();
                }
            }
        }.runTaskLaterAsynchronously(Core.INSTANCE, 1200);
    }

    /**
     * Remove the invitee from the invitees
     *
     * @param invitee
     */
    public void removeInvite(Player invitee) {
        invited.remove(invitee.getUniqueId());
    }

    /**
     * Check whether a player is invited
     *
     * @param invitee
     * @return
     */
    public boolean isInvited(Player invitee) {
        return invited.contains(invitee.getUniqueId());
    }

    /**
     * Change the leader of the party
     *
     * @param leader
     */
    public void setLeader(UUID leader) {
        if (this.leader != null) {
            Core.partyManager.getParties().remove(this.leader);
            this.addMember(this.leader);
        }
        this.leader = leader;
        if (members.contains(leader)) this.removeMember(leader);
        Core.partyManager.getParties().put(leader, this);
        save();
    }

    /**
     * Disband the party
     */
    public void disband() {
        Core.partyManager.getParties().remove(leader);
        this.members.clear();
        this.invited.clear();
        // Core.plugin().partyConfig.getConfiguration().set("party." + partyID.toString(), null); TODO: Uncomment when parties
    }

    /**
     * Leave the party
     *
     * @param player
     */
    public void leave(Player player) {
        if (isMember(player.getUniqueId())) removeMember(player.getUniqueId());
        if (getLeader() == player) disband();
        send(player.getName() + " left the party!");
    }

    /**
     * Send a message to all party members
     *
     * @param message
     */
    public void send(String message) {
        for (OfflinePlayer player : getMembers()) {
            if (player.isOnline()) {
                ((Player) player).sendMessage("§b[PARTY] §7" + message);
            }
        }
    }

    /**
     * Save all members to the config
     *
     * @param save Whether to save the config or keep it in memory only
     */
    public void save(boolean save) {
        List<String> membersString = new ArrayList<>();
//        FileConfiguration config = Core.plugin().partyConfig.getConfiguration();
//        members.forEach((UUID u) -> membersString.add(u.toString()));
//        config.set("party." + partyID.toString() + ".id", getPartyID().toString());
//        config.set("party." + partyID.toString() + ".leader", leader.toString());
//        config.set("party." + partyID.toString() + ".members", membersString);
//        if (save)
//            try {Core.plugin().partyConfig.save(); } catch(Exception ignored){} TODO: Uncomment when parties
    }

    /**
     * Save all members to the config (in memory)
     */
    public void save() {
        save(false);
    }
}

File: core/src/main/kotlin/net/siegerpg/siege/core/parties/Party.kt
package net.siegerpg.siege.core.parties

import net.kyori.adventure.text.Component
import net.kyori.adventure.text.minimessage.MiniMessage
import net.kyori.adventure.text.minimessage.Template
import net.siegerpg.siege.core.Core
import net.siegerpg.siege.core.miscellaneous.Utils
import org.bukkit.Bukkit
import org.bukkit.OfflinePlayer
import org.bukkit.configuration.ConfigurationSection
import org.bukkit.entity.Player
import java.time.Instant
import java.util.*

class Party(public val partyID : UUID, private var leader : OfflinePlayer) {

	private val invited : ArrayList<Player> = ArrayList()
	private val members : HashSet<OfflinePlayer> = HashSet()

	private var lastJoin : Instant = Instant.now()

	constructor(leader : OfflinePlayer) : this(UUID.randomUUID(), leader)

	init {
		parties[partyID] = this
	}

	fun getLeader() : OfflinePlayer {
		return leader
	}

	fun addMember(member : OfflinePlayer) {
		members.add(member)
		lastJoin = Instant.now()
		save()
	}

	fun removeMember(member : OfflinePlayer) {
		members.remove(member)
		save()
		Bukkit.getScheduler().runTaskLater(Core.plugin(), Runnable {
			if (getMembers().size < 2 && lastJoin.plusSeconds(60L * 5)
							.isBefore(Instant.now())
			) {
				send(Utils.tacc("&cThe party was disbanded because nobody was in it for 5 minutes!"))
				disband()
			}
		}, 20L * 60 * 5)
	}

	fun getMembers() : HashSet<OfflinePlayer> {
		val guys : HashSet<OfflinePlayer> = HashSet()
		guys.addAll(this.members)
		guys.add(getLeader())
		return guys
	}

	fun isMember(player : OfflinePlayer) : Boolean {
		return getMembers().contains(player)
	}

	fun addInvite(invited : Player) {
		this.invited.add(invited)
		Bukkit.getScheduler().runTaskLaterAsynchronously(
				Core.plugin(),
				Runnable {
					if (!this.invited.contains(invited)) return@Runnable
					send(
							MiniMessage.get()
									.parse(
											Utils.tacc("&7The party invite to &6<invitedDisplayName> &r&7expired!"),
											listOf(
													Template.of(
															"invitedDisplayName",
															invited.displayName()
													           )
											      )
									      ),

							)
					val leader = if (getLeader().isOnline) getLeader() as Player else null
					val leaderName = leader?.displayName() ?: getLeader().name
					invited.sendMessage(Utils.tacc("&7Party invite from &6${leaderName} &r&7has expired!"))
				}, (2 * 60 * 20).toLong()
		                                                )
	}

	fun removeInvite(invited : Player) {
		this.invited.remove(invited)
	}

	fun isInvited(invited : Player) : Boolean {
		return this.invited.contains(invited)
	}

	fun setLeader(newLeader : Player) {
		if (!isMember(newLeader)) return
		addMember(leader)
		removeMember(newLeader)
		leader = newLeader
		save()
	}

	fun disband() {
		send(Utils.tacc("&7The party has been disbanded!"))
		delete()
	}

	fun delete() {
		members.clear()
		invited.clear()
		parties.remove(partyID)
		Core.plugin().partyConfig.setParty(partyID, null)
	}

	fun leave(player : Player) {
		if (getLeader() == player) disband()
		else if (isMember(player)) {
			send(
					MiniMessage.get().parse(
							"<gold><displayName> <reset><gray>has left the party!",
							listOf(
									Template.of("displayName", player.displayName())
							      )
					                       )
			    )
			removeMember(player)
		}
	}

	fun send(message : Component) {
		for (member in getMembers()) {
			if (member.isOnline) {
				val onlineMember = member as Player
				onlineMember.sendMessage(message)
			}
		}
	}

	fun send(message : String) {
		for (member in getMembers()) {
			if (member.isOnline) {
				val onlineMember = member as Player
				onlineMember.sendMessage(message)
			}
		}
	}

	fun kick(player : Player) {
		if (player == getLeader()) {
			send(Utils.tacc("&6${leader.name} &r&7can't be kicked from the party!"))
			return
		}
		send(
				MiniMessage.get()
						.parse(
								"<gold>${leader.name} <gray>kicked <gold><kickedDisplayName> <reset><gray>from the party!",
								listOf(
										Template.of(
												"kickedDisplayName",
												player.displayName()
										           )
								      )
						      ),
		    )
		removeMember(player)
	}


	companion object {

		val parties : HashMap<UUID, Party> = HashMap()

		fun deserialize(uuid : UUID, section : ConfigurationSection) : Party {
			val leaderID = UUID.fromString(section.getString("leader"))
			val leader = Bukkit.getOfflinePlayer(leaderID)
			val party = Party(uuid, leader)
			section.getStringList("members").forEach { member ->
				val memberID = UUID.fromString(member)
				val memberOfflinePlayer = Bukkit.getOfflinePlayer(memberID)
				party.addMember(memberOfflinePlayer)
			}

			return party
		}

		fun getPlayerParty(player : OfflinePlayer) : Party? {
			for ((_, party) in parties) {
				if (party.isMember(player))
					return party
			}
			return null
		}

		fun getPlayerInvites(player : Player) : HashMap<UUID, Party> {
			val partiesWithInvites = hashMapOf<UUID, Party>()
			for ((partyID, party) in parties) {
				if (party.isInvited(player)) {
					partiesWithInvites[partyID] = party
				}
			}
			return partiesWithInvites
		}
	}

	private fun save() {
		val partySection = Core.plugin().partyConfig.getParty(partyID)
		partySection.set(
				"members",
				members.map { m -> m.uniqueId.toString() }.toTypedArray()
		                )
		partySection.set("leader", getLeader().uniqueId.toString())
		Core.plugin().partyConfig.save()
	}
}